### PR TITLE
IEP-463 DFU (USB) programming support

### PR DESCRIPTION
**Pull request by sigmaaa**
PR summary:
- added a DFU toggle button to activate  `idf.py -p /dev/ttyACM0 dfu-flash` and `dfu` command for build
- warning message, when the selected target does not support DFU

Need to add:
- new icon 
- documentation update

### CODE REVIEW

**Comment by sigmaaa** [comment]
Documentation is updated now

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/DfuCommandsUtil.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+				return windows[0].getShell();
+			}
+		}
+		else
+		{
+			return window.getShell();
+		}
+		return null;
+	}
+}
```
Comment (alirana01) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/DfuCommandsUtil.java:133]: Class needs to be formatted properly

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java [hunk 2]
```diff
@@ -88,6 +89,16 @@ public void launch(ILaunchConfiguration configuration, String mode, ILaunch laun
...
+		serialPort = ((SerialFlashLaunch) launch).getLaunchTarget()
+				.getAttribute(SerialFlashLaunchTargetProvider.ATTR_SERIAL_PORT, ""); //$NON-NLS-1$
+
+		if (DfuCommandsUtil.isDfu()) {
+			if (checkIfPortIsEmpty(configuration)) {
+				return;
+			}
+			DfuCommandsUtil.flashDfuBins(getProject(configuration), launch, monitor, serialPort);
+			return;
+		}
```
Comment (alirana01) [inline bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java:101]: Code not formatted properly here as well

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/DfuCommandHandler.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	     Command command = event.getCommand();
+	     HandlerUtil.toggleCommandState(command);
+		return event;
+	}
+
+	@Override
+	protected void execute() {
+	}
+	
+}
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/DfuCommandHandler.java:27]: Formatting not proper

**Review by alirana01 — COMMENTED** [review]
LGTM as per code review needs functional testing properly though and can we add some kind of functional test for it without using any hardware? If we need hardware even it would be good if you can just write this in a new test class in a separate branch and we will figure out later and connect some boards to the runner to run such tests as well, this is not highly required at the moment though but would be good if you can do it in some other branch.

**Comment by alirana01** [comment]
There are also some conflicts on this which needs to be resolved as well with rebase of master on the branch and be careful with the properties file as this may mess up the properties previously added

**Comment by sigmaaa** [comment]
> 

Hi Ali, thanks for the review. We can add a functional test for it without hardware, for example, we can check if DFU binaries are generated after the build command. I'll add it

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/DfuCommandsUtil.java (hunk 1 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/DfuCommandsUtil.java]: fixed

File: bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java (hunk 2 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.core/src/com/espressif/idf/launch/serial/internal/SerialFlashLaunchConfigDelegate.java]: fixed

**Comment by sigmaaa** [comment]
added dfu build test case

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/DfuCommandHandler.java (hunk 3 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/DfuCommandHandler.java]: fixed

**Comment by kolipakakondal** [comment]
LGTM @sigmaaa Could you resolve the conflicts and merge this PR. Please make sure you choose "Squash and Merge"

**Pull request merged**